### ISSUE TITLE
HasUrlComponent#setParameter is invoked clicking on self-referencing links

### ISSUE THREAD

**Issue by heruan** [vaadin/flow#4649]
The `setParameter` method of `HasUrlComponent` gets invoked even if the user is already in there; this means all the operations the method does with the parameter are executed again. For example:

```java
public class MainLayout extends VerticalLayout implements RouterLayout {

    public MainLayout() {
        add(new RouterLink("One", Foo.class, 1);
        add(new RouterLink("Two", Foo.class, 2);
        add(new RouterLink("Three", Foo.class, 3);
    }
}

@Route(value = "foo", layout = MainLayout.class)
public class FooCard extends VerticalLayout implements HasUrlParamenter<Long> {

    @Override
    public void setParameter(BeforeEvent event, Long parameter) {
        composeLayoutFor(parameter);
    }
}
```

If the user is at `/foo/1` and clicks on link "One", `composeLayoutFor` is invoked even if the location does not change, for example duplicating components that the method might add (which depends on the parameter value).

**Comment by Legioth** [comment]
This is supposed to be handled by `Router.handleNavigationForLocation`,  but it's apparently broken in some way.

**Comment by knoobie** [comment]
Sounds like you should involve somebody from the UX Team as well. On most websites, if you click a link again the site refreshes. Don't know what you want to arrive with flow.

Example: Clicking two times on 'Issues' here in github.

**Comment by Legioth** [comment]
That's a good point. By calling `setParameter` again, the application developer can choose whether they want to ignore the event or refresh something. If not calling `setParameter`, the application developer wouldn't have any way of triggering a refresh.

**Comment by heruan** [comment]
I agree with @knoobie: clicking on a self-referencing link should act the same as when clicking another in-app link: a new instance is created and `setParameter` is invoked on that instance.

**Comment by heruan** [comment]
Another variant of this: navigating from `/foo/1` to `/foo/2`, I would expect a new instance of `Foo` to be created, replacing the existing instance; instead, `setParameter` is invoked in the *same* instance. @Legioth, is this the intended behavior?

**Comment by heruan** [comment]
I've replicated what I consider the issue: [SelfReferencingLink.java](https://github.com/heruan/vaadin-spring-demo/blob/44a59431060002614c9ca5718280bce8533381fe/src/main/java/to/lova/vaadin/spring/SelfReferencingLink.java)

1. Run with `mvn spring-boot:run` and go to http://localhost:8080/self-referencing-link/foo
2. Click on [Foo](http://localhost:8080/self-referencing-link/foo) or [Bar](http://localhost:8080/self-referencing-link/bar): the instance is always the same, which is a problem is there is some state.

Not sure if this is intended or not, i.e. if views are supposed not to have a state.

**Comment by heruan** [comment]
Seems like the culprit is at [AbstractNavigationStateRenderer.java#110](https://github.com/vaadin/flow/blob/40a89b9998c35073db06f4a1073bb56b84835027/flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java#L110):

```java
    static <T extends HasElement> T getRouteTarget(Class<T> routeTargetType,
            NavigationEvent event) {
        UI ui = event.getUI();
        Optional<HasElement> currentInstance = ui.getInternals()
                .getActiveRouterTargetsChain().stream()
                .filter(component -> component.getClass()
                        .equals(routeTargetType))
                .findAny();
        return (T) currentInstance.orElseGet(() -> Instantiator.get(ui)
                .createRouteTarget(routeTargetType, event));
    }
```

The component instance is taken from the active ones just checking its class, ignoring the cases where there is a parameter that might have changed.

I see two scenarios:

1. the user is in `/foo/1` and clicks on `/foo/2`: the target instance *must be new*, as if the it were a different route targets;
2. the user is in `/foo/1` and clicks on `/foo/1`: what to do in this case?
    1. the same as in [1], i.e. new target instance (as @knoobie suggested)
    2. keep the same instance and state (as it is now, which I stronlgy discourage)
    3. do nothing, i.e. ignore the click.

In my opinion both [2.i] and [2.iii] have their own pros and cons, but I guess [2.i] is what the user expects to happen.

**Comment by knoobie** [comment]
I think it depends on the "scope". Do we have a view / component that is in View or UI scope? 

-  I would expect that a view / component in **ViewScope** gets re-created everytime it gets visited. 
-  Otherwise for a view / component in **UIScope** I would expect that it's reused. 

Nevertheless 'clicking on the same Link again' should trigger everything like "F5" does. That's "normal" user behaviour on a lot of sites.

**Comment by pleku** [comment]
>Another variant of this: navigating from /foo/1 to /foo/2, I would expect a new instance of Foo to be created, replacing the existing instance (in-app, without refresh); instead, setParameter is invoked in the same instance.

I can confirm that this is the intented behavior. The Foo should just update its state based on the new state (1->2), instead of throwing everything away and creating it again.

If you want to create everything again, and are eg. using Spring, it should be very easy to achieve that too. But the default is intended to be like it is now.

**Comment by heruan** [comment]
Thank you @pleku for the feedback! I understand the default is optimal when there is a simple state to update, but it could be cumbersome to replace an existing complex state; e.g. if you have components which are attached depending on the parameter, you should track them to be able to remove them later.

Do you think it would be acceptable to have an option in `@Route` to set this behavior?

> using Spring, it should be very easy to achieve that

Maybe I'm missing something obvious, but… what are you referring to? 🙂

**Comment by Legioth** [comment]
Making it configurable e.g. through the `@Route` annotation sounds like a good way to go. I guess the API would then be something like `@Route(value = "foo", reuseInstance = false)`. Should preferably also take router layouts into account since those are also reused using the same mechanism.

I think the comment about Spring can be disregarded. The original router prototype implementation delegated the decision whether to reuse an instance or create a new one to Spring, but the current implementation always reuses based on the code snippet referenced in the [above comment](https://github.com/vaadin/flow/issues/4649#issuecomment-437567812).

**Comment by pleku** [comment]
>but the current implementation always reuses based on the code snippet referenced in the above comment.
Should probably also consider in https://github.com/vaadin/spring/issues/263 that it should be possible to always create a new view instance on navigation for Spring too. Maybe when `prototype` scope is used ?
